### DISCUSSION TITLE
A build for the browser

### DISCUSSION THREAD

**Discussion by mercmobily** [eslint/eslint#15733] [Ideas]
I need to run Eslint in the browser. I am going totally insane trying to do it.
I realise you don't want to support the browser environment. However, I can't work out how to create a browser-consumable bundle with all of the rules baked in, which I can then load from the browser (specifically, for CodeMirror 6).

This repository https://github.com/satya164/eslint-browser is some 2400 commits behind (last commit, 5 years ago). I am thinking I could try browserify.

Before I start with more head banging, can I ask... How do you guys make the bundle for the demo exactly?

And... are you sure it's not worthwhile to at least provide some guidance for this?

**Comment by nzakas** [comment]
We don’t want to support a browser build because it will require a lot more maintenance on our part. ESLint was never intended to run in the browser but we are kind of hacking it. We can’t guarantee that the way we do it now will continue to work in the long run.

But in general, you should be able to Webpack the `Linter` class and have it work in a browser:

https://github.com/eslint/eslint/blob/main/webpack.config.js

**Reply by jsudelko** [reply]
Done right, supporting the browser should reduce maintenance. There are a lot of competing platforms: Node, Deno, Bun, Cloudflare and friends where people may conceivably want to run their code (and probably one or two more within the decade). These platforms are now converging towards supporting browser APIs and as such supporting the browser as a first class citizen improves portability and decreases maintenance.

Even for tools mostly used outside the browser, there's increasing consensus (particularly for newer JS tooling) that the way to go is to have the core runnable entirely in the browser, and adapters for other platforms such as Node, Deno etc. These could be supported by the ESLint team or the community as need arises.

**Reply by nzakas** [reply]
Already being discussed. See https://github.com/eslint/eslint/discussions/16557.

**Comment by mercmobily** [comment]
Is that webpack config what you use at the moment for your demo?

**Reply by nzakas** [reply]
That one is used for running some of our tests in a browser. I'm not 100% sure how the demo works at this point. You should check out the source code: https://github.com/eslint/website/tree/master/src/js/demo

**Reply by mercmobily** [reply]
I really am trying to figure it out, but I am having a lot of trouble. I am looking at the demo, and the whole site generator https://github.com/eslint/website I am having a great deal of trouble figuring out where Eslint actually gets built.

I don't suppose I could humbly ask you for some direction?

Reality is, people are using Elint on browsers. A lot. There is a lot of online editing at the moment. The tools to translate things to the browser are much better than they were.
I think you should at least provide _some_ support  -- even if it's limited, even if it's guidance -- to run Eslint on the browser. There is so much head-banging around at the moment...

**Reply by dvhx** [reply]
@mercmobily Did you managed to run eslint in browser? I also need browser version because node version has million dependencies and can write/erase anything on the disk while browser version is nicely isolated and cannot do any harm.

**Reply by mdjermanovic** [reply]
You can also check the new playground repository:

https://github.com/eslint/playground

**Reply by mercmobily** [reply]
I ended up using https://www.npmjs.com/package/eslint4b-prebuilt which uses https://www.npmjs.com/package/eslint4b -- yeah, I know...
eslint4b seems to do some funky magic to make sure that things work in the browser.

HOWEVER.

The guys at Eslint have finally released the playground https://github.com/eslint/playground -- I don't know React, but I looked at the code, and they seem to simply load it:

````
import { Linter, SourceCodeFixer } from "./node_modules/eslint/lib/linter/";
````

So, I am not sure what the fuss is about, in terms of 'needing' eslint4b. But then again, I couldn't even work out how the rules were loaded. So...

**Reply by aladdin-add** [reply]
yes, the `"eslint/lib/linter"` should just work. it won't include node's builtin-modules like fs.

Note it's not in the public exports, so it can be changed in future releases (though not likely happen very often 😄 ). I'd suggest you use an exact version, or a lockfile like `package-lock.json`.

**Reply by mercmobily** [reply]
@aladdin-add Does Eslint itself (that is, NOT eslitn-cli) ever, ever need "fs"? I mean, somebody wrote a whole script that is meant to get rid of all node dependencies from Eslitn... but at this point, I wonder: is this even needed at all?

**Reply by aladdin-add** [reply]
yes, it needs "fs" to read configs/plugins.

**Reply by mercmobily** [reply]
OK, but the linter itself doesn't right? (Above, I meant to write "Does the Linter itself") For example your "playground" repository just loads the linter: https://github.com/eslint/playground/blob/main/src/playground/App.js So, as long as you don't load configs nor plugins, you don't need `fs` right?

**Reply by aladdin-add** [reply]
yes,exactly!

**Reply by mercmobily** [reply]
Sorry to pester you -- and thank you for donating your time to me (I am normally on the other end). I understand with configs -- you need to load them, obviously. But, why do you need `fs` for plugins? Aren't they just `import()`ed? Or, `require()`ed?

**Reply by mercmobily** [reply]
(I am assuming it would ever make sense to load a plugin in a browser context... does it?)

**Reply by aladdin-add** [reply]
as it needs to know to load which plugins (`"plugins"` were configured in the eslintrc). So, with the linter, if you want to use rules in a 3rd-party plugin, you'll get an error like "Definition for rules 'xxx' was not found."

**Reply by mercmobily** [reply]
I understand. But aren't plugins loaded via a dynamic `require()`? Why `fs` in that case? (This is my last question regardless -- thank you)

**Reply by aladdin-add** [reply]
emm..maybe you are right. but it did need some node.js builtin-modules, so you cannot use it directly on browsers.
(I'm on a mobile, so cannot search where it was used 😄 )

**Reply by mercmobily** [reply]
No worries at all. Thank you!
